pytorch_lightning/trainer/trainer.py

```diff
@@ -791,41 +847,63 @@ def __test_using_best_weights(self, ckpt_path, test_dataloaders):
             model.load_state_dict(ckpt['state_dict'])
 
         # attach dataloaders
-        if test_dataloaders is not None:
-            self.data_connector.attach_dataloaders(model, test_dataloaders=test_dataloaders)
+        if dataloaders is not None:
+            loaders_arg = {
+                'validation': 'val_dataloaders',
+                'test': 'test_dataloaders'
+            }[stage]
+
+            kwargs = {
+                loaders_arg: dataloaders
+            }
+
+            self.data_connector.attach_dataloaders(model, **kwargs)
 
         # run tests
+        self.evaluating = stage
         self.tested_ckpt_path = ckpt_path
         self.testing = True
         os.environ['PL_TESTING_MODE'] = '1'
         self.model = model
         results = self.fit(model)
-        self.testing = False
         del os.environ['PL_TESTING_MODE']
+        self.testing = False
+        self.evaluating = False
 
         # teardown
         if self.is_function_implemented('teardown'):
             model_ref = self.get_model()
-            model_ref.teardown('test')
+            model_ref.teardown(stage)
 
         return results
 
-    def __test_given_model(self, model, test_dataloaders):
+    def __evaluate_given_model(self, model, dataloaders, stage: str):
 
         # attach data
-        if test_dataloaders is not None:
-            self.data_connector.attach_dataloaders(model, test_dataloaders=test_dataloaders)
+        if dataloaders is not None:
+            loaders_arg = {
+                'validation': 'val_dataloaders',
+                'test': 'test_dataloaders'
+            }[stage]
+
+            kwargs = {
+                loaders_arg: dataloaders
+            }
+
+            self.data_connector.attach_dataloaders(model, **kwargs)
 
         # run test
         # sets up testing so we short circuit to eval
-        self.testing = True
+        self.evaluating = stage
+        self.testing = True  # TODO: remove, keep only evaluating
```

Comment:
My idea is that long-term we could deprecate `testing` and only keep `evaluating = 'validation'` or `evaluating = 'test'`.

It's not clear to me if it's intended to be public API or it's just an internal flag that we could remove right away.

Reply:
What is your opinion here? @rohitgr7 @carmocca

Reply:
So currently, `stage` can be `fit|test`
where `fit` actually implies `train + validation`

It makes sense to me to have `stage` be `fit|validation|test`

Reply:
Here I would:
* Either remove `testing` or transform it to a read-only computed property computed from `evaluating`, which raises a DeprecationWarning if used; 
* Remove `PL_TESTING_MODE`;
* Maybe rename `tested_ckpt_path` to `evaluated_ckpt_path`

Reply:
+1 `It makes sense to me to have stage be fit|validation|test`

> Maybe rename tested_ckpt_path to evaluated_ckpt_path

I don't see why `tested_ckpt_path` is there. It's not used anywhere.

> Remove `PL_TESTING_MODE`

This too seems redundant, not sure why it's there. Making a wild guess that this is meant to be used when user manually sets `PL_TESTING_MODE = 1` and run `trainer.fit()`, in such case it should do testing.

> Either remove testing or transform it to a read-only property computed from evaluating, which raises a DeprecationWarning if used;

+1 :heavy_check_mark:

Reply:
> I don't see why tested_ckpt_path is there. It's not used anywhere.

I saw it used in the tests, I think it's meant to allow the caller of `trainer.test()` to determine which checkpoint was loaded.

> This too seems redundant, not sure why it's there. Making a wild guess that this is meant to be used when user manually sets PL_TESTING_MODE = 1 and run trainer.fit(), in such case it should do testing.

I went through the commit history and the commit that first added it only read it in DDP mode: [here](https://github.com/PyTorchLightning/pytorch-lightning/commit/f35337adba0557dddc05bd6b10b896ba68a8e4cf#diff-667a2513c158c2b01d73b479e4dea75587d8531f9cd286891d34570a2fd145dcR973). After a refactor which moved all accelerator-specific code out of trainer.py, that line was left there but is now executed in every case.

I don't know enough about DDP to know why this matters. Maybe if you set the `trainer.testing` attribute in the parent process it's not accessible by the children, so and environment variable was used insted? Perhaps we can ask the author @williamFalcon to chime in?

> Either remove testing or transform it to a read-only property computed from evaluating, which raises a DeprecationWarning if used;

@rohitgr7 do you suggest deprecating or remove right away?

Reply:
I'd suggest deprecate it for now. If it needs to be removed later sometime in this PR then we can do that too easily.